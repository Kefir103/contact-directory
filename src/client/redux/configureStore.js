import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/index';
import { rootReducer } from './reducers/rootReducer';
import thunk from 'redux-thunk';

let store;

export default function configureStore(initialState) {
    if (store) {
        return store;
    }

    store =
        process.env.NODE_ENV === 'production'
            ? createStore(rootReducer, initialState, compose(applyMiddleware(thunk)))
            : createStore(rootReducer, initialState, composeWithDevTools(applyMiddleware(thunk)));

    return store;
}
