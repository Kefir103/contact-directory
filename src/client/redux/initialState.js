export const initialState = {
    filter: {
        sortingMap: new Map([]),
        filterText: '',
        filterFields: [],
        currentPage: 1,
    },
    data: {
        currentElement: {
            elementInfo: undefined,
            elementIndex: -1,
        },
        elements: [],
        currentElements: [],
        filteredElements: undefined,
        inputContainer: {
            isAddButtonDisabled: false,
            isFormOpen: false,
            inputElement: {
                id: 0,
                firstName: '',
                lastName: '',
                email: '',
                phone: '',
            },
            validInputs: {
                id: false,
                firstName: false,
                lastName: false,
                email: false,
                phone: false,
            },
            addingStatus: {
                addingStatusText: '',
                isError: false,
            },
        },
    },
    appStatus: {
        isLoading: false,
        error: undefined,
    },
};
