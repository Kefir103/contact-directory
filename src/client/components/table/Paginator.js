import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setCurrentPage } from '../../redux/actions/filterActions';

const Paginator = (props) => {
    const isCurrentPage = (currentPage, buttonText) => {
        return currentPage === buttonText ? 'selected' : '';
    };

    const handleFirstPageButtonClick = () => {
        props.actions.setCurrentPage(1);
    };

    const handlePageButtonClick = (event) => {
        props.actions.setCurrentPage(Number(event.target.value));
    };

    const handleLastPageButtonClick = () => {
        props.actions.setCurrentPage(props.pageCount);
    };

    const renderButtons = (pageCount, currentPage) => {
        const buttons = [];
        for (let i = 1, iCounter = 0; i <= pageCount && iCounter < 5; i++) {
            if (Math.abs(i - currentPage) <= 2) {
                buttons.push(
                    <button
                        className={`page-button ${isCurrentPage(currentPage, i)}`}
                        onClick={handlePageButtonClick}
                        value={i}
                        key={`pageBtn${i}`}>
                        {i}
                    </button>
                );
                iCounter++;
            }
        }
        return buttons;
    };

    return (
        <div className={'paginator'}>
            <button
                className={'page-button arrow'}
                onClick={handleFirstPageButtonClick}
                style={
                    props.currentPage - 3 > 0 && props.pageCount > 5
                        ? { visibility: 'visible' }
                        : { visibility: 'hidden' }
                }>
                &lt;&lt;
            </button>
            {renderButtons(props.pageCount, props.currentPage)}
            <button
                className={'page-button arrow'}
                onClick={handleLastPageButtonClick}
                style={
                    props.pageCount - props.currentPage > 2 && props.pageCount > 5
                        ? { visibility: 'visible' }
                        : { visibility: 'hidden' }
                }>
                &gt;&gt;
            </button>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        currentPage: state.filter.currentPage,
        pageCount: state.data.pageCount,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        actions: bindActionCreators(
            {
                setCurrentPage,
            },
            dispatch
        ),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Paginator);
