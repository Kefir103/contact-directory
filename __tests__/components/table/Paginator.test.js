import React from 'react';
import { shallow, render } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import Paginator from '../../../src/client/components/table/Paginator';
import { initialState } from '../../../src/client/redux/initialState';

const mockStore = configureMockStore();

describe('<Paginator />', () => {
    test('should render correctly at the start of paginating', () => {
        const store = mockStore({
            ...initialState,
            filter: { ...initialState.filter, currentPage: 1 },
            data: { ...initialState.data, pageCount: 10 },
        });
        const wrapper = shallow(<Paginator store={store} />);
        const component = wrapper.dive().dive();
        expect(component).toMatchSnapshot();
    });
    test('should render correctly at the end of paginating', () => {
        const store = mockStore({
            ...initialState,
            filter: { ...initialState.filter, currentPage: 10 },
            data: { ...initialState.data, pageCount: 10 },
        });
        const wrapper = shallow(<Paginator store={store} />);
        const component = wrapper.dive().dive();
        expect(component).toMatchSnapshot();
    });
    test('should render correctly in the middle of paginating', () => {
        const store = mockStore({
            ...initialState,
            filter: { ...initialState.filter, currentPage: 5 },
            data: { ...initialState.data, pageCount: 10 },
        });
        const wrapper = shallow(<Paginator store={store} />);
        const component = wrapper.dive().dive();
        expect(component).toMatchSnapshot();
    });
    test('should render correctly in the middle of paginating but with pageCount = 5', () => {
        const store = mockStore({
            ...initialState,
            filter: { ...initialState.filter, currentPage: 3 },
            data: { ...initialState.data, pageCount: 5 },
        });
        const wrapper = shallow(<Paginator store={store} />);
        const component = wrapper.dive().dive();
        expect(component).toMatchSnapshot();
    });
});
