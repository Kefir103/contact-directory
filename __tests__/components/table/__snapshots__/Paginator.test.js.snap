// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> should render correctly at the end of paginating 1`] = `
<div
  className="paginator"
>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "visible",
      }
    }
  >
    &lt;&lt;
  </button>
  <button
    className="page-button "
    key="pageBtn8"
    onClick={[Function]}
    value={8}
  >
    8
  </button>
  <button
    className="page-button "
    key="pageBtn9"
    onClick={[Function]}
    value={9}
  >
    9
  </button>
  <button
    className="page-button selected"
    key="pageBtn10"
    onClick={[Function]}
    value={10}
  >
    10
  </button>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    &gt;&gt;
  </button>
</div>
`;

exports[`<Paginator /> should render correctly at the start of paginating 1`] = `
<div
  className="paginator"
>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    &lt;&lt;
  </button>
  <button
    className="page-button selected"
    key="pageBtn1"
    onClick={[Function]}
    value={1}
  >
    1
  </button>
  <button
    className="page-button "
    key="pageBtn2"
    onClick={[Function]}
    value={2}
  >
    2
  </button>
  <button
    className="page-button "
    key="pageBtn3"
    onClick={[Function]}
    value={3}
  >
    3
  </button>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "visible",
      }
    }
  >
    &gt;&gt;
  </button>
</div>
`;

exports[`<Paginator /> should render correctly in the middle of paginating 1`] = `
<div
  className="paginator"
>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "visible",
      }
    }
  >
    &lt;&lt;
  </button>
  <button
    className="page-button "
    key="pageBtn3"
    onClick={[Function]}
    value={3}
  >
    3
  </button>
  <button
    className="page-button "
    key="pageBtn4"
    onClick={[Function]}
    value={4}
  >
    4
  </button>
  <button
    className="page-button selected"
    key="pageBtn5"
    onClick={[Function]}
    value={5}
  >
    5
  </button>
  <button
    className="page-button "
    key="pageBtn6"
    onClick={[Function]}
    value={6}
  >
    6
  </button>
  <button
    className="page-button "
    key="pageBtn7"
    onClick={[Function]}
    value={7}
  >
    7
  </button>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "visible",
      }
    }
  >
    &gt;&gt;
  </button>
</div>
`;

exports[`<Paginator /> should render correctly in the middle of paginating but with pageCount = 5 1`] = `
<div
  className="paginator"
>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    &lt;&lt;
  </button>
  <button
    className="page-button "
    key="pageBtn1"
    onClick={[Function]}
    value={1}
  >
    1
  </button>
  <button
    className="page-button "
    key="pageBtn2"
    onClick={[Function]}
    value={2}
  >
    2
  </button>
  <button
    className="page-button selected"
    key="pageBtn3"
    onClick={[Function]}
    value={3}
  >
    3
  </button>
  <button
    className="page-button "
    key="pageBtn4"
    onClick={[Function]}
    value={4}
  >
    4
  </button>
  <button
    className="page-button "
    key="pageBtn5"
    onClick={[Function]}
    value={5}
  >
    5
  </button>
  <button
    className="page-button arrow"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    &gt;&gt;
  </button>
</div>
`;
